File: napakalaki/src/Napakalaki/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Napakalaki;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author marcos
 */
public class Napakalaki {
    
    private Player currentPlayer;
    private ArrayList<Player> players;
    private CardDealer dealer;
    private Monster currentMonster;
    private static final Napakalaki INSTANCE = new Napakalaki();
    
    private Napakalaki() {
        players = new ArrayList();
        currentMonster= null;
        currentPlayer = null;
        dealer=CardDealer.getInstance();
    }
    
    public static Napakalaki getInstance() {
        return INSTANCE;
    }
       
    private void initPlayers( ArrayList<String> names){ 
        players.add(new PlayerBlindado(names.get(0)));
        for (int i=1; i<names.size(); i++) {
            players.add(new Player(names.get(i)));
        }
        
    }
    
    private Player nextPlayer(){
        Player aux;
        Random generator=new Random();
        if(currentPlayer==null){
                aux= players.get(generator.nextInt(players.size()));
        }
        else{
            if(players.get(players.size()-1)==currentPlayer){
                aux= players.get(0);
            }
            else{
                int indice=players.indexOf(currentPlayer);
                aux = players.get(indice+1);
            }
        }
        return aux;
    }
    
    private boolean nextTurnAllowed(){
        if(currentPlayer==null){
            return true;
        }
        else{
            return currentPlayer.validState();
        }
    }
    
    private void setEnemies(){
        Random generator=new Random();
        int numeroJugadores=players.size();
        int indice;
            for(int i=0; i<numeroJugadores; i++){
                indice = (generator.nextInt(numeroJugadores));
                if(i != indice){
                  players.get(i).setEnemy(players.get(indice));
                }
                else{
                    i--;
                }
        }
        
    }
    
    public CombatResult developCombat(){
        CombatResult aux = currentPlayer.combat(currentMonster);
        if(aux==CombatResult.LOSEANDCONVERT){
            Cultist c=dealer.nextCultist();
            CultistPlayer cp=new CultistPlayer(currentPlayer,c);
            int indice=players.indexOf(currentPlayer);
            players.set(indice, cp);
            
            for(Player p: players){
                if(p.getEnemy()==currentPlayer){
                    p.setEnemy(cp);
                }
            }
            currentPlayer=cp;
        }
        dealer.giveMonsterBack(currentMonster);
        return aux;
    }
    
    public void discardVisibleTreasure(ArrayList<Treasure> treasures){
        for(Treasure t: treasures){
            currentPlayer.discardVisibleTreasure(t);
            dealer.giveTreasureBack(t);
        }
    }
    
    public void discardHiddenTreasure(ArrayList<Treasure> treasures){
        for(Treasure t: treasures){
            currentPlayer.discardHiddenTreasure(t);
            dealer.giveTreasureBack(t);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(Treasure t:treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> names){
        this.initPlayers(names);
        this.setEnemies();
        dealer.initCards();
        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK, dead;
        stateOK=this.nextTurnAllowed();
        if(stateOK){
            currentMonster=dealer.nextMonster();
            currentPlayer=this.nextPlayer();
            dead=currentPlayer.isDead();
            if(dead){
              currentPlayer.initTreasures();
            }
        }
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        return result==CombatResult.WINGAME;
    }

    public String toString(){
        return "Jugador Actual = " + currentPlayer
                + " Jugadores = " + players
                + " Mazo de cartas = " + dealer
                + " Monstruo actual = " + currentMonster
                + "\n";
    }
    
    
}

File: napakalaki(ruby)/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame
require 'singleton'
require_relative 'player'
require_relative 'combat_result'
require_relative 'monster'
require_relative 'card_dealer'
require_relative 'cultist_player'


class Napakalaki
  include Singleton
  attr_reader :currentPlayer, :currentMonster
  
  def initialize
    @currentPlayer=nil
    @players=Array.new
    @dealer=CardDealer.instance
    @currentMonster=nil
   
  end
  
  def initPlayers(names)
    names.each do |n|
      @players.push(Player.new(n))
    end
  end
  
  def nextPlayer
    tamanio=@players.length
    indice=rand(tamanio)
    if(@currentPlayer==nil)
      aux= @players[indice]
    else
      if(@players[tamanio-1]==@currentPlayer)
        aux= @players[0]
      else
        indice=@players.index(@currentPlayer)
        aux = @players[indice+1];
      end
    end
    return aux;
  end
  
  def nextTurnAllowed
    if(@currentPlayer==nil)
      return true
    else
      return @currentPlayer.validState
    end  
  end
  
  def setEnemies
    numero_jugadores=@players.length
        i=0
        while i < numero_jugadores
          indice = (rand(numero_jugadores))
            if(i != indice)
              @players[i].enemy=@players[indice]
            else
              i=i-1
            end
            i=i+1
        end
  end
  
  def developCombat
    aux = @currentPlayer.combat(@currentMonster)
    if(aux==CombatResult::LOSEANDCONVERT)
      c=@dealer.nextCultist
      cul=CultistPlayer.new(@currentPlayer, c)
      posicion=@players.index(@currentPlayer)
      @currentPlayer=cul
        @players.each do |p|
          if(p.enemy==@currentPlayer)
            p.enemy=cul
          end
        end
      @players[posicion]=cul
    end
    @dealer.giveMonsterBack(@currentMonster)
    return aux
  end
  
  def discardVisibleTreasures(treasures)
    treasures.each do |t|
      @currentPlayer.discardVisibleTreasure(t)
      @dealer.giveTreasureBack(t)
    end
  end
  
  def discardHiddenTreasures(treasures)
    treasures.each do |t|
      @currentPlayer.discardHiddenTreasure(t)
      @dealer.giveTreasureBack(t)
    end
  end
  
  def makeTreasuresVisible(treasures)
    treasures.each do |t|
      @currentPlayer.makeTreasureVisible(t)
    end
  end
  
  def initGame(names)
    initPlayers(names)
    setEnemies()
    @dealer = CardDealer.instance()
    @dealer.initCards
    nextTurn()
  end
  
  #def getCurrentPlayer
    #return @currentPlayer
  #end
  
  #def getCurrentMonster
    #return @currentMonster
  #end
  
  def nextTurn
    state_ok=nextTurnAllowed
    if(state_ok)
      @currentMonster=@dealer.nextMonster
      @currentPlayer=nextPlayer
      dead=@currentPlayer.dead
      if(dead)
        @currentPlayer.initTreasures
      end    
    end
    return state_ok;
  end
  
  def endOfGame(result)
    if result == CombatResult::WINGAME
      return true
    else
      return false
    end
  end
  
  def to_s
   "\nActual Jugador : #{@currentPlayer} \n Jugadores: #{@players} \n Actual Monstruo: #{@currentMonster}"
  end
  
  private :initPlayers, :nextPlayer, :nextTurnAllowed, :setEnemies
  
end
end